### PR TITLE
Configurable cipher suites

### PR DESCRIPTION
**Pull request by beavenj**
### Type of change

- Enhancement / new feature

Contributes to: #2643 

Signed-off-by: John Beaven beavenj@uk.ibm.com

### Description

Added the ssl.cipher.suites Kafka configuration parameter to the list of allowed parameters (ssl. is excluded at the present time). This allowed the user to lock down the allowable cipher suites that Kafka will support.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [x] Write tests
- [x] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Reference relevant issue(s) and close them after merging
- [x] Update CHANGELOG.md

### CODE REVIEW

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/AbstractConfigurationTest.java [hunk 1]
```diff
@@ -238,6 +238,16 @@ public void testKafkaZookeeperTimeout() {
...
     }
 
+    @Test
+    public void testKafkaCipherSuiteOverride() {
+        Map<String, Object> conf = new HashMap<>();
+        conf.put("ssl.cipher.suites", "cipher1,cipher2,cipher3"); // valid
+
+        KafkaConfiguration kc = new KafkaConfiguration(conf.entrySet());
+
+        assertThat(kc.asOrderedProperties().asMap().get("ssl.cipher.suites"), is("cipher1,cipher2,cipher3"));
```
Comment (samuel-hawker) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/AbstractConfigurationTest.java]: Hi John, from a test debugging perspective, the below might be a little bit cleaner:
```
assertThat(kc.asOrderedProperties().asMap(), hasEntry("ssl.cipher.suites", "cipher1,cipher2,cipher3"));
```
Where you statically import hasEntry with: `import static org.hamcrest.Matchers.hasEntry;`

**Comment by tombentley** [comment]
> * How are the users expected to know which cipher suites are supported? Configuring wrong cipher suite would mean basically breaking the whole cluster.

I agree that letting the user configure the wrong thing, e.g. because of a typo, isn't great UX. However, it would be difficult to do completely reliably, I think, because the CO doesn't know what suites are installed in the broker (or connect etc) image. Do we know whether and how the broker dies when the cipher suite is a wrong value?

**Comment by ppatierno** [comment]
I can understand and totally agree on the bad UX but do you have something different for other parameters in the `config` section? I don't think so, right? Today if user puts an invalid value for a field, just screw up the broker on start up. It would happen the same for wrong ssl cipher suites.

**Comment by tombentley** [comment]
@ppatierno I think we do do some validation, using the "config model" stuff. Of course this is never perfect, but it will catch things like passing a list where a number is expected. Precisely because it will never be perfect we will need to draw the line somewhere and maybe put some effort into getting a good message back to the user after the fact (e.g. via the pod termination log), rather than trying to get perfect validation before.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/AbstractConfigurationTest.java (hunk 1 above)
Comment (beavenj) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/model/AbstractConfigurationTest.java]: Hi Sam, this code is consistent with the other tests on which is was based, so I would suggest this change be left as is and then if its desirable to modify them, we do all of them together in another PR. Just my thoughts on that one...

Comment (samuel-hawker) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/model/AbstractConfigurationTest.java]: Perfectly reasonable John!

**Comment by beavenj** [comment]
Hi, I will double check exactly what gets logged in the event that there no cipher suite can be used to see if its clear or not what the cause of the issue is...

**Comment by beavenj** [comment]
@tombentley, @ppatierno Hi, I double checked the behaviour when a cipher suite that Kafka does not support is specified in this setting. If any of the cipher suites is not supported or recognised then an error similar to this one is generated:

```
2020-03-23 15:39:50,856 ERROR [KafkaServer id=1] Fatal error during KafkaServer startup. Prepare to shutdown (kafka.server.KafkaServer) [main]
org.apache.kafka.common.KafkaException: org.apache.kafka.common.config.ConfigException: Invalid value java.lang.IllegalArgumentException: Unsupported ciphersuite TLS_AES_128_GCM_SHA256 for configuration A client SSLEngine created with the provided settings can't connect to a server SSLEngine created with those settings.
	at org.apache.kafka.common.network.SslChannelBuilder.configure(SslChannelBuilder.java:71)
	at org.apache.kafka.common.network.ChannelBuilders.create(ChannelBuilders.java:146)
	at org.apache.kafka.common.network.ChannelBuilders.serverChannelBuilder(ChannelBuilders.java:85)
	at kafka.network.Processor.<init>(SocketServer.scala:753)
	at kafka.network.SocketServer.newProcessor(SocketServer.scala:394)
	at kafka.network.SocketServer.$anonfun$addDataPlaneProcessors$1(SocketServer.scala:279)
	at scala.collection.immutable.Range.foreach$mVc$sp(Range.scala:158)
	at kafka.network.SocketServer.addDataPlaneProcessors(SocketServer.scala:278)
	at kafka.network.SocketServer.$anonfun$createDataPlaneAcceptorsAndProcessors$1(SocketServer.scala:241)
	at kafka.network.SocketServer.$anonfun$createDataPlaneAcceptorsAndProcessors$1$adapted(SocketServer.scala:238)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at kafka.network.SocketServer.createDataPlaneAcceptorsAndProcessors(SocketServer.scala:238)
	at kafka.network.SocketServer.startup(SocketServer.scala:121)
	at kafka.server.KafkaServer.startup(KafkaServer.scala:263)
	at kafka.server.KafkaServerStartable.startup(KafkaServerStartable.scala:44)
	at kafka.Kafka$.main(Kafka.scala:84)
	at kafka.Kafka.main(Kafka.scala)
```

In this case its because this suite requires TLS 1.3.. This happens even if there are valid cipher suites in the list. The effect is to enforce that all specified suites are valid. So if the user includes one thats either not supported or is mis-spelt it will be easily detected when the Kafka brokers fails to start by looking at their logs.

**Comment by scholzj** [comment]
Ok, so lets say we cannot do much better that this and letting it to the user to see the error and fix it. What about the other points I raised?

> Are you planning to add the same also for other compoenents? Connect? Mirror Maker? Mirror Maker 2? So that users can use this from all compoenents.

and 

>I also wonder if we should do the same with other SSL options as well - such as the ssl.enabled.protocols for example (see #2406). If we let users configure the cipher suites, we can probably let them configure the protocols as well, or?

Should we leave it to other PR or do it here?

**Comment by beavenj** [comment]
@scholzj Hi, since these are separate components I would propose we do them under their own PRs. To that end I have raised issues #2772 for KafkaConnect and #2774 for MirrorMaker and MirrorMaker2.. Can we merge this PR and we will then address the others as well...

Regards,
John.

**Comment by scholzj** [comment]
@ppatierno @tombentley Are you fine with this approach? Thanks.

@beavenj Could you please fix the conflicts so that I can run some more tests on this PR? Thanks.

**Comment by tombentley** [comment]
@scholzj fine with me

**Comment by scholzj** [comment]
@strimzi-ci run tests

**Comment by scholzj** [comment]
@strimzi-ci run tests

**Comment by scholzj** [comment]
@strimzi-ci run tests profile=acceptance testcase=io.strimzi.systemtest.ConnectST#testKafkaConnectAndConnectorFileSinkPlugin

**Comment by scholzj** [comment]
Thanks for the PR.

**Comment by beavenj** [comment]
Your welcome

**Pull request merged**